File: Weekly Contest 235 (0403)/3/kmjp.java
// 04/04/21 night

import java.util.*;
import static java.lang.System.out;

public class kmjp {
	private final int MAX = Integer.MAX_VALUE;
	private final int MIN = Integer.MIN_VALUE;
	private final int MOD = 1000000007;

    // Accepted --- 188ms 60.00%
	public int minAbsoluteSumDiff(int[] a, int[] b) {
		TreeSet<Integer> se = new TreeSet<>(Arrays.asList(MAX, MIN));
		long res, sum;
		res = sum = 0;
		int n = a.length;
		for (int i = 0; i < n; i++) {
			se.add(a[i]);
			sum += Math.abs(a[i] - b[i]);
		}
		// out.println(se);
		res = sum;
		for (int i = 0; i < n; i++) {
			int it = lower_bound(se, b[i]);
			// out.println(it);
			res = Math.min(res, sum - Math.abs(a[i] - b[i]) + Math.abs(it - b[i]));
			it = se.lower(it);
			// out.println(it);
			res = Math.min(res, sum - Math.abs(a[i] - b[i]) + Math.abs(it - b[i]));
		}
		return (int) (res % MOD);
	}

	public int lower_bound(TreeSet<Integer> se, int t) {
		int res = -1;
		if (se.contains(t)) {
			res = t;
		} else {
			res = se.higher(t);
		}
		return res;
	};

	public static void main(String[] args) {
		int [] nums1 = {1, 7, 5};
		int [] nums2 = {2, 3, 5};
		int [] nums1_2 = {2, 4, 6, 8, 10};
		int [] nums2_2 = {2, 4, 6, 8, 10};
		int [] nums1_3 = {1, 10, 4, 4, 2, 7};
		int [] nums2_3 = {9, 3, 5, 1, 7, 4};
		kmjp t = new kmjp();
		// out.println(t.minAbsoluteSumDiff(nums1, nums2));
		// out.println(t.minAbsoluteSumDiff(nums1_2, nums2_2));
		// out.println(t.minAbsoluteSumDiff(nums1_3, nums2_3));

		// Testts
		TreeSet<Integer> ts = new TreeSet<>(Arrays.asList(1, 7, 5));
		out.println(ts.lower(-5));
        out.println(ts.lower(0));
		out.println(ts.lower(1));
		out.println(ts.lower(2));
		out.println(ts.lower(3));
		out.println(ts.lower(4));
		out.println(ts.lower(5));
		out.println(ts.lower(6));
		out.println(ts.lower(7));
		out.println(ts.lower(9));
		out.println(ts.lower(1000));
		out.println();
		out.println(ts.higher(-5));
		out.println(ts.higher(0));
		out.println(ts.higher(1));
		out.println(ts.higher(2));
		out.println(ts.higher(3));
		out.println(ts.higher(4));
		out.println(ts.higher(5));
		out.println(ts.higher(6));
		out.println(ts.higher(7));
		out.println(ts.higher(9));
		out.println(ts.higher(1000));
	}
}

File: Contest/2021/Weekly Contest 240 (0508)/2/kmjp.js
// 05/08/21 night
const pr = console.log;


// Accepted --- 100ms
const maxDistance = (a, b) => {
    let n = a.length;
    let m = b.length;
    let res = 0;
    let j = 0;
    for (let i = 0; i < n; i++) {
        for (j = mx(i, j); j < m && a[i] <= b[j]; j++);
        res = mx(res, j - i - 1);
    }
    return res;
};

// Accepted --- 104ms
const mx = Math.max;
const maxDistance1 = (a, b) => {
    let n = a.length;
    let m = b.length;
    let j = -1;
    let res = 0;
    for (let i = 0; i < n; i++) {
        j = mx(i, j);
        // j = i; // Accepted --- 3696ms
        while (j + 1 < m && a[i] <= b[j + 1]) j++;
        // pr(i, j, a[i], b[j], j - i);
        res = mx(res, j - i);
    }
    return res;
};

const main = () => {
    let nums1 = [55, 30, 5, 4, 2], nums2 = [100, 20, 10, 10, 5];
    let nums1_2 = [2, 2, 2], nums2_2 = [10, 10, 1];
    let nums1_3 = [30, 29, 19, 5], nums2_3 = [25, 25, 25, 25, 25];
    let nums1_4 = [5, 4], nums2_4 = [3, 2];
    let nums1_debug1 = [55, 30, 5, 4, 2], nums2_debug1 = [100, 20, 10, 10, 5];
    let nums1_debug2 = [2], nums2_debug2 = [1];
    pr(maxDistance(nums1, nums2));
    pr(maxDistance(nums1_2, nums2_2));
    pr(maxDistance(nums1_3, nums2_3));
    pr(maxDistance(nums1_4, nums2_4));
    pr(maxDistance(nums1_debug1, nums2_debug1));
    pr(maxDistance(nums1_debug2, nums2_debug2));
};

main()